File: server/src/main/scala/com/lightbend/modelServer/processor/VotingDecider.scala
package com.lightbend.modelServer.processor

import com.lightbend.modelServer.model.ServingResult
import com.lightbend.modelServer.model.speculative.{CurrentProcessing, Decider}


object VotingDecider extends Decider with Serializable{

  // The simple voting decider for results 0 or 1. Returning 0 or 1
  override def decideResult(results: CurrentProcessing): ServingResult = {

    var result = ServingResult.noModel
    var sum = .0
    var count = 0
    var source = 0
    results.results.foreach(res => res.result match {
      case r if(r.processed) =>
        sum = sum + r.result.getOrElse(0).asInstanceOf[Int]
        if(r.source != source) source = r.source
        count = count + 1
      case _ =>
    })
    if(count == 0) result
    else {
      val res = if(sum/count < .5) 0 else 1
      ServingResult("voter model", source, Some(res), System.currentTimeMillis() - results.start)
    }
  }
}

File: server/src/main/java/com/lightbend/java/modelServer/processor/VotingDecider.java
package com.lightbend.java.modelServer.processor;

import com.lightbend.java.modelServer.model.ServingResult;
import com.lightbend.java.modelServer.model.speculative.CurrentProcessing;
import com.lightbend.java.modelServer.model.speculative.Decider;
import com.lightbend.java.modelServer.model.speculative.ServingResponse;

import java.util.Optional;

public class VotingDecider implements Decider {

    @Override public ServingResult decideResult(CurrentProcessing results) {
        ServingResult result = ServingResult.getEmpty();
        double sum = .0;
        int count = 0;
        int source = 0;
        for (ServingResponse res : results.getResults()) {
            if (res.getResult().isProcessed()) {
                sum += (int) (res.getResult().getResult().orElse(0));
                if (res.getResult().getSource() != source) source = res.getResult().getSource();
                count += 1;
            }
        }
        if (count == 0)
            return result;
        int res = (sum / count < .5) ? 0 : 1;
        return new ServingResult("voter model", source, Optional.of(res), System.currentTimeMillis() - results.getStart());
    }
}
